Fix/audit

daodesigner opened the pull request.

ctrlc03 committed:
fix(grantround): reverting `transferMatchingFunds` when funds in the contract are not enough

Added a `revert` statement in the `transferMatchingFunds` function to prevent the function returning
without having send the funds to the recipient. This checks that the balance of the contract is >=
`_payoutAmount`. Also added a unit test.

#163

ctrlc03 committed:
fix(qfi and grantround): implemented fixes for all issues related to token transfers safety

This commit fixes:
 1. Not reverting in `transferMatchingFunds` (`GrantRound`) if funds are not
enough.
 2. Using arbitrary tokens with the contract. Now all related functions make sure that the
balance of the receiver equals to the balance before the transfer + the amout transferred.

fix #161 fix #163 fix #164 fix #157

ctrlc03 committed:
fix(qfi): implemented fixes to prevent possible reentrancy attacks

Moved up the change of `STAGE` for the functions that could be affected by reentrancy such as
`finalizeCurrentRound`, `initialize`, `contribute`, and `deployGrantRound`. Also changed one error
message to reflect the new reason for revertin in `initialize`.

fix #162

ctrlc03 committed:
fix(qfi): mismatched arrays length and incorrect error messages fix

Implemented fixes to prevent users from submitting mismatched arrays of pubKeys and messages on
`publishMessageBatches`. Also adjusted the error messages in certain `revert` statements
(`closeVotingAndWaitForDeadline` and `acceptContributionsAndTopUpsBeforeNewRound`) to reflect the
actual stage of the contract. Appropriate test cases were added too.

fix #158, fix #159

ctrlc03 committed:
fix(qfi and grantround): documentation upgrade and changes to GrantRound's `transferMatchingFunds`

Added missing documentation as well as upgraded certain descriptions to reflect the new changes of
the contracts. This documents the risks of using untrusted tokens and tokens with blacklisting
capabilities. Furthermore, the parameter `_erc20Address` within `transferMatchingFunds`
(`GrantRound`) was removed and the `nativeToken` is used instead. Furthermore, the contributor count
stored in the `QFI` contract is now reduced after a user withdraws their contribution.

fix #160, fix #161

ctrlc03 committed:
fix(fundsmanager): implemented a check to prevent users from removing arbitrary funding sources

Implemented some logic to prevent any user from removing a funding source in FundsManager. Now, only
the admin can remove arbitrary funding sources, wheres a funding source can remove themselves. Also
amended certain error messages to reflect they are coming from the FundsManager contract.

ctrlc03 committed:
fix(qfi): implemented contributions and withdrawals per round

Implemented changes to allow contributions to multiple rounds. Now, whenever a user contributes, the
amount is stored in relation to the current GrantRound. After deploying a new GrantRound, their
contributions will not count anymore as these would have already been distributed.

fix #165

ctrlc03 committed:
feat(qfi): feat to keep track of contributions per round

Implented a feat to keep track of the contributions per each round (which increase when calling
`contribute` and decrease when calling `withdraw`) and a view function to count the total
contributions of all rounds.

daodesigner commented on .husky/pre-commit:
```diff
@@ -1,37 +1,6 @@
 
+BASE_DIR="$( cd "$( dirname "$0" )" && pwd )"
 
-# Return `true` if files under `packages/contracts/contracts/` directory has been changed or added. Otherwise, return `false`
```
keep this section

ctrlc03 replied:
This is taken from here: https://github.com/quadratic-funding/qfi/commit/af01221850a843e33f3c7f171c0ae637b58d8a15 - what should I change? @daodesigner

daodesigner replied:
oo nvm then

daodesigner reviewed (commented):
Changes to contracts are neat and and appreciate you taking care of nits like _ for internal functions, commenting the code, and updating the natspec

Missing integration tests (please do them on a separate file, ie. tests/QF/09-audit)

ctrlc03 committed:
test(contracts): implemented integration test

First batch of integration tests for the full project. Includes testing of `QFI`, `GrantRound`,
`GrantRoundFactory`, `FundsManager`, `OptimisticRecipientRegistry` and `BaseRecipientRegistry`.

ctrlc03 committed:
test(qfi and grantround): implemented more tests for GrantRound and QFI

Started writing tests related to the finalization of the process.

ctrlc03 committed:
test(qfi and grantround): implemented more integration tests on the QFI side

Completed tests related to the finalization and transfer of matching funds from QFI to GrantRound so
that they can be paid out accordingly.

gurrpi commented:
🚀

ctrlc03 committed:
docs(qfi): updated docs to reflect audit fix

Updated QFI docs to reflect a change in the `finalizeCurrentRound` function where two redundant
arguments were removed.

ctrlc03 committed:
Merge branch 'feat/code-freeze' into fix/audit

ctrlc03 committed:
test(qfi): amended test that fail due to the timestamp

Amended two test cases which sometimes fail due to the block timestamp not matching.

ctrlc03 committed:
Merge branch 'fix/audit' of github.com:quadratic-funding/qfi into fix/audit

ctrlc03 committed:
test(qfi): amended another test which fails sometimes due to the block timestamp

Amended `getVoiceCredits` tests on `QFI` due to random failure on mismatched block timestamp values
